### ISSUE TITLE
exporting data from netlogo run

### ISSUE THREAD

**Issue by kurt-klingbeil** [NetLogo/NetLogo#1357]
secondary observation related to
https://github.com/NetLogo/NetLogo/issues/1356

In poking around looking for data/evidence, we created separate wealth graphs for the 4 turtles
wealth0-3
after a run, we exported the data and then did some additional excel monkeying around as attached

the most striking observation is that the data record is bizarre
A reasonable expectation is that at _every_ step there is a single fixed-sum wealth transfer from
one randomly-selected turtle to another one
Instead, one observes that there are records involving 0,1,2 wealth transfers (in addition to the
anomalous non-fixed-sum nature of the transfers)

Any brilliant ideas/comprehensive explanations as to what is going on ? 
I would call it buggy behaviour. 
Perhaps someone can cogently shatter my expectation of 
[SimpEcon_WDsn_VariousTaxScheme_2017Feb28 TotalWealth.xlsx](https://github.com/NetLogo/NetLogo/files/822158/SimpEcon_WDsn_VariousTaxScheme_2017Feb28.TotalWealth.xlsx)
single-transaction per single-step records ...

**Comment by mrerrormessage** [comment]
Posted originally [on issue 1356](https://github.com/NetLogo/NetLogo/issues/1356#issuecomment-284765777), reproduced here for completeness. The below answer assumes this data corresponds to the code in #1356 .

NetLogo's ask and one-of primitives appear to be similar to while-loops and first in other languages, but they're actually quite different. ask (and other agentset primitives) evaluate the block they are passed for each agent in a random (but reproducible) order. Similarly one-of returns a random (but reproducible) agent from the agentset it takes as an argument. This is actually a major component of NetLogo's design. From the programming guide:

>  An agentset is not in any particular order. In fact, it's always in a random order. And every time you use it, the agentset is in a different random order. This helps you keep your model from treating any particular turtles, patches or links differently from any others (unless you want them to be). Since the order is random every time, no one agent always gets to go first.

Because the agentset primitives are random, there are a combinatorial number of paths that this code might take (as there often are when working with agentsets). I'll enumerate these inductively starting with zero turtles and moving to two

1. There are no turtles with wealth > tax: nothing happens.
1. There is exactly one turtle with wealth > tax: the block is run for that turtle, it decreases its own wealth and picks exactly one other turtle at random and increases that turtle's wealth.
1. There are two turtles with wealth > tax: the block is run for each turtle, the following possible outcomes can occur based on the selection of one-of.
    1. Both turtles select a third and fourth turtle. Both of the turtles have their wealth decreased by their tax and the third and fourth turtles have their wealth increased: four turtles change their wealth.
      1. The first and second turtle both select the same third turtle. The first and second turtles have their wealth decreased by their tax and the third turtle has its wealth increased: three turtles change their wealth.
      1. One turtle selects the other turtle while the second selects some third turtle. Each of these turtles has its wealth increased. In the first version, the first turtle will have its wealth decreased by its tax, the third turtle will have its wealth increased by its own tax, and the second turtle will have its wealth unchanged: only two turtles change their wealth. In the second version, the first turtle will have its wealth decreased by its own tax, the second turtle will have its wealth changed by the difference between the first turtle's tax and its own tax, and the third turtle will have its wealth increased by the second turtles tax: only three turtles change their wealth (and only two if the two turtles start with the same wealth).
      1. The turtles select each other and increase the other's wealth. In the first version there are zero turtles which change their wealth. In second their are either two or zero turtles that change their wealth depending on whether or not the two turtles have the same wealth.

Because the possibilities are combinatoric, you will get even more possible outcomes when moving beyond two turtles with wealth > tax (and even wider swings in the number of turtles which appear to have changed their wealth, especially under the first procedure).

In summary, I believe that NetLogo is behaving as intended here and I will be closing this issue. If you believe that we've misunderstood something in your question, please let us know!
